7

Return a table where all the customers who order the Product with ID 12 are returned with their Company Name

<span style="color: var(--vscode-foreground);">The order by here is optional, I put it in to make it look like the desired output in the exercise</span>

In [1]:
USE Northwinds2022TSQLV7;

SELECT C.CustomerID, C.CustomerCompanyName
 FROM Sales.Customer as C
 WHERE EXISTS (
	SELECT *
	FROM Sales.[Order] as O
	WHERE O.CustomerId = C.CustomerId
	AND EXISTS (
		SELECT * 
		FROM Sales.OrderDetail as OD
		WHERE OD.OrderId = O.OrderId
		AND OD.ProductId = 12
	)
 )
 ORDER BY C.CustomerCompanyName;

(12 rows affected)

Total execution time: 00:00:00.059

CustomerID CustomerCompanyName 48 Customer DVFMB 39 Customer GLLAG 71 Customer LCOUJ 65 Customer NYUHS 44 Customer OXFRU 51 Customer PVDZC 86 Customer SNXOJ 20 Customer THHDP 90 Customer XBBVR 46 Customer XPNIK 31 Customer YJCBX 87 Customer ZHYOS

8 (Optional, Advanced)

Write a query that calculates a running total qty

for each customer and month using subqueries

Tables involved: TSQLV4 database, Sales.CustOrders view

DROP TABLE IF EXISTS Sales.CustOrder;

In [3]:
USE Northwinds2022TSQLV7


SELECT O1.CustomerId, CONVERT(VARCHAR(7), O1.OrderDate, 120) as OrderMonth, SUM(OD.Quantity) as QTY,
 (SELECT SUM(OD2.Quantity)
  FROM 
        Sales.OrderDetail AS OD2
        INNER JOIN Sales.[Order] as O2 ON OD2.OrderId = O2.OrderId
    WHERE 
        O2.CustomerId = O1.CustomerId
        AND CONVERT(VARCHAR(7), O2.OrderDate, 120) <= CONVERT(VARCHAR(7), O1.OrderDate, 120)
 ) AS runqty
FROM Sales.[Order] AS O1
INNER JOIN Sales.OrderDetail as OD
	ON OD.OrderId = O1.OrderId
GROUP BY O1.CustomerID, CONVERT(VARCHAR(7), O1.OrderDate, 120)
ORDER BY O1.CustomerId, CONVERT(VARCHAR(7), O1.OrderDate, 120);

Use TSQLV4

SELECT custid, ordermonth, qty,
(SELECT SUM(O2.qty)    
FROM Sales.CustOrders AS O2
WHERE O2.custid = O1.custid 
AND O2.ordermonth <= O1.ordermonth
) AS runqty
FROM Sales.CustOrders AS O1 
ORDER BY custid, ordermonth;

(636 rows affected)

(636 rows affected)

Total execution time: 00:00:00.225

CustomerId OrderMonth QTY runqty 1 2015-08 38 38 1 2015-10 41 79 1 2016-01 17 96 1 2016-03 18 114 1 2016-04 60 174 2 2014-09 6 6 2 2015-08 18 24 2 2015-11 10 34 2 2016-03 29 63 3 2014-11 24 24 3 2015-04 30 54 3 2015-05 80 134 3 2015-06 83 217 3 2015-09 102 319 3 2016-01 40 359 4 2014-11 50 50 4 2014-12 55 105 4 2015-02 40 145 4 2015-06 96 241 4 2015-10 89 330 4 2015-11 43 373 4 2015-12 103 476 4 2016-02 19 495 4 2016-03 124 619 4 2016-04 31 650 5 2014-08 126 126 5 2014-12 43 169 5 2015-02 84 253 5 2015-05 87 340 5 2015-06 87 427 5 2015-08 52 479 5 2015-09 65 544 5 2015-10 35 579 5 2015-11 61 640 5 2015-12 10 650 5 2016-01 167 817 5 2016-02 128 945 5 2016-03 56 1001 6 2015-04 23 23 6 2015-06 18 41 6 2015-07 27 68 6 2016-01 10 78 6 2016-03 34 112 6 2016-04 28 140 7 2014-07 50 50 7 2014-09 80 130 7 2014-11 138 268 7 2015-02 168 436 7 2015-06 143 579 7 2015-08 25 604 7 2015-09 12 616 7 2016-01 50 666 8 2014-10 90 90 8 2015-12 60 150 8 2016-03 40 190 9 2014-10 87 87 9 2014-11 94 181 9 2015-03 53 234 9 2015-04 110 344 9 2015-05 45 389 9 2015-09 80 469 9 2015-10 51 520 9 2015-11 147 667 9 2016-01 36 703 9 2016-02 119 822 9 2016-03 108 930 9 2016-05 50 980 10 2014-12 81 81 10 2015-01 269 350 10 2015-04 80 430 10 2015-11 105 535 10 2016-03 289 824 10 2016-04 132 956 11 2014-08 39 39 11 2015-03 77 116 11 2015-05 52 168 11 2015-06 26 194 11 2015-07 10 204 11 2016-03 55 259 11 2016-04 34 293 12 2015-04 19 19 12 2015-12 1 20 12 2016-01 27 47 12 2016-02 10 57 12 2016-03 28 85 12 2016-04 30 115 13 2014-07 11 11 14 2014-07 57 57 14 2014-12 65 122 14 2015-04 66 188 14 2015-11 153 341 14 2016-03 32 373 14 2016-04 92 465 15 2014-08 60 60 15 2015-03 15 75 15 2015-04 30 105 15 2016-03 9 114 15 2016-04 19 133 16 2015-02 32 32 16 2015-03 22 54 16 2016-01 33 87 17 2014-11 44 44 17 2014-12 18 62 17 2015-12 20 82 17 2016-01 32 114 17 2016-04 37 151 17 2016-05 9 160 18 2014-09 13 13 18 2015-07 19 32 18 2015-09 9 41 18 2016-02 39 80 19 2014-11 35 35 19 2015-01 86 121 19 2015-05 39 160 19 2015-11 30 190 19 2016-03 86 276 19 2016-04 293 569 20 2014-07 305 305 20 2014-11 198 503 20 2014-12 334 837 20 2015-01 411 1248 20 2015-02 170 1418 20 2015-04 233 1651 20 2015-06 39 1690 20 2015-07 215 1905 20 2015-08 164 2069 20 2015-09 59 2128 20 2015-10 108 2236 20 2015-12 551 2787 20 2016-01 337 3124 20 2016-02 346 3470 20 2016-03 271 3741 20 2016-04 602 4343 20 2016-05 200 4543 21 2014-11 70 70 21 2014-12 25 95 21 2015-01 68 163 21 2015-04 37 200 21 2015-06 50 250 21 2015-08 85 335 21 2015-10 22 357 23 2015-01 51 51 23 2015-03 42 93 23 2015-08 102 195 23 2015-12 159 354 24 2014-07 60 60 24 2014-10 140 200 24 2014-12 6 206 24 2015-02 49 255 24 2015-05 98 353 24 2015-06 60 413 24 2015-10 75 488 24 2015-12 186 674 24 2016-01 21 695 24 2016-02 146 841 24 2016-03 142 983 24 2016-04 251 1234 25 2014-07 135 135 25 2014-10 297 432 25 2014-12 121 553 25 2015-03 50 603 25 2015-06 35 638 25 2015-08 94 732 25 2015-09 312 1044 25 2015-10 72 1116 25 2015-12 34 1150 25 2016-01 105 1255 25 2016-03 124 1379 25 2016-04 146 1525 26 2015-09 32 32 26 2016-01 23 55 26 2016-03 14 69 27 2015-01 2 2 27 2015-10 10 12 27 2015-11 9 21 27 2015-12 1 22 27 2016-04 32 54 28 2014-10 59 59 28 2014-11 30 89 28 2015-03 91 180 28 2015-05 100 280 28 2015-07 16 296 28 2015-09 51 347 28 2016-03 2 349 29 2014-11 10 10 29 2015-01 12 22 29 2015-06 5 27 29 2016-02 5 32 29 2016-03 10 42 30 2014-09 85 85 30 2015-05 34 119 30 2015-08 29 148 30 2016-02 151 299 30 2016-03 53 352 30 2016-04 43 395 31 2015-01 34 34 31 2015-09 61 95 31 2015-10 78 173 31 2015-11 65 238 31 2015-12 43 281 31 2016-03 20 301 31 2016-04 14 315 32 2015-05 20 20 32 2015-07 93 113 32 2015-09 107 220 32 2016-01 50 270 32 2016-03 30 300 32 2016-04 45 345 33 2014-07 14 14 33 2015-12 20 34 34 2014-07 162 162 34 2015-05 84 246 34 2015-08 35 281 34 2015-10 50 331 34 2015-12 30 361 34 2016-02 226 587 34 2016-03 147 734 34 2016-04 105 839 35 2014-07 46 46 35 2014-12 106 152 35 2015-03 186 338 35 2015-04 49 387 35 2015-05 48 435 35 2015-07 143 578 35

custid ordermonth qty runqty 1 2015-08-01 38 38 1 2015-10-01 41 79 1 2016-01-01 17 96 1 2016-03-01 18 114 1 2016-04-01 60 174 2 2014-09-01 6 6 2 2015-08-01 18 24 2 2015-11-01 10 34 2 2016-03-01 29 63 3 2014-11-01 24 24 3 2015-04-01 30 54 3 2015-05-01 80 134 3 2015-06-01 83 217 3 2015-09-01 102 319 3 2016-01-01 40 359 4 2014-11-01 50 50 4 2014-12-01 55 105 4 2015-02-01 40 145 4 2015-06-01 96 241 4 2015-10-01 89 330 4 2015-11-01 43 373 4 2015-12-01 103 476 4 2016-02-01 19 495 4 2016-03-01 124 619 4 2016-04-01 31 650 5 2014-08-01 126 126 5 2014-12-01 43 169 5 2015-02-01 84 253 5 2015-05-01 87 340 5 2015-06-01 87 427 5 2015-08-01 52 479 5 2015-09-01 65 544 5 2015-10-01 35 579 5 2015-11-01 61 640 5 2015-12-01 10 650 5 2016-01-01 167 817 5 2016-02-01 128 945 5 2016-03-01 56 1001 6 2015-04-01 23 23 6 2015-06-01 18 41 6 2015-07-01 27 68 6 2016-01-01 10 78 6 2016-03-01 34 112 6 2016-04-01 28 140 7 2014-07-01 50 50 7 2014-09-01 80 130 7 2014-11-01 138 268 7 2015-02-01 168 436 7 2015-06-01 143 579 7 2015-08-01 25 604 7 2015-09-01 12 616 7 2016-01-01 50 666 8 2014-10-01 90 90 8 2015-12-01 60 150 8 2016-03-01 40 190 9 2014-10-01 87 87 9 2014-11-01 94 181 9 2015-03-01 53 234 9 2015-04-01 110 344 9 2015-05-01 45 389 9 2015-09-01 80 469 9 2015-10-01 51 520 9 2015-11-01 147 667 9 2016-01-01 36 703 9 2016-02-01 119 822 9 2016-03-01 108 930 9 2016-05-01 50 980 10 2014-12-01 81 81 10 2015-01-01 269 350 10 2015-04-01 80 430 10 2015-11-01 105 535 10 2016-03-01 289 824 10 2016-04-01 132 956 11 2014-08-01 39 39 11 2015-03-01 77 116 11 2015-05-01 52 168 11 2015-06-01 26 194 11 2015-07-01 10 204 11 2016-03-01 55 259 11 2016-04-01 34 293 12 2015-04-01 19 19 12 2015-12-01 1 20 12 2016-01-01 27 47 12 2016-02-01 10 57 12 2016-03-01 28 85 12 2016-04-01 30 115 13 2014-07-01 11 11 14 2014-07-01 57 57 14 2014-12-01 65 122 14 2015-04-01 66 188 14 2015-11-01 153 341 14 2016-03-01 32 373 14 2016-04-01 92 465 15 2014-08-01 60 60 15 2015-03-01 15 75 15 2015-04-01 30 105 15 2016-03-01 9 114 15 2016-04-01 19 133 16 2015-02-01 32 32 16 2015-03-01 22 54 16 2016-01-01 33 87 17 2014-11-01 44 44 17 2014-12-01 18 62 17 2015-12-01 20 82 17 2016-01-01 32 114 17 2016-04-01 37 151 17 2016-05-01 9 160 18 2014-09-01 13 13 18 2015-07-01 19 32 18 2015-09-01 9 41 18 2016-02-01 39 80 19 2014-11-01 35 35 19 2015-01-01 86 121 19 2015-05-01 39 160 19 2015-11-01 30 190 19 2016-03-01 86 276 19 2016-04-01 293 569 20 2014-07-01 305 305 20 2014-11-01 198 503 20 2014-12-01 334 837 20 2015-01-01 411 1248 20 2015-02-01 170 1418 20 2015-04-01 233 1651 20 2015-06-01 39 1690 20 2015-07-01 215 1905 20 2015-08-01 164 2069 20 2015-09-01 59 2128 20 2015-10-01 108 2236 20 2015-12-01 551 2787 20 2016-01-01 337 3124 20 2016-02-01 346 3470 20 2016-03-01 271 3741 20 2016-04-01 602 4343 20 2016-05-01 200 4543 21 2014-11-01 70 70 21 2014-12-01 25 95 21 2015-01-01 68 163 21 2015-04-01 37 200 21 2015-06-01 50 250 21 2015-08-01 85 335 21 2015-10-01 22 357 23 2015-01-01 51 51 23 2015-03-01 42 93 23 2015-08-01 102 195 23 2015-12-01 159 354 24 2014-07-01 60 60 24 2014-10-01 140 200 24 2014-12-01 6 206 24 2015-02-01 49 255 24 2015-05-01 98 353 24 2015-06-01 60 413 24 2015-10-01 75 488 24 2015-12-01 186 674 24 2016-01-01 21 695 24 2016-02-01 146 841 24 2016-03-01 142 983 24 2016-04-01 251 1234 25 2014-07-01 135 135 25 2014-10-01 297 432 25 2014-12-01 121 553 25 2015-03-01 50 603 25 2015-06-01 35 638 25 2015-08-01 94 732 25 2015-09-01 312 1044 25 2015-10-01 72 1116 25 2015-12-01 34 1150 25 2016-01-01 105 1255 25 2016-03-01 124 1379 25 2016-04-01 146 1525 26 2015-09-01 32 32 26 2016-01-01 23 55 26 2016-03-01 14 69 27 2015-01-01 2 2 27 2015-10-01 10 12 27 2015-11-01 9 21 27 2015-12-01 1 22 27 2016-04-01 32 54 28 2014-10-01 59 59 28 2014-11-01 30 89 28 2015-03-01 91 180 28 2015-05-01 100 280 28 2015-07-01 16 296 28 2015-09-01 51 347 28 2016-03-01 2 349 29 2014-11-01 10 10 29 2015-01-01 12 22 29 2015-06-01 5 27 29 2016-02-01 5 32 29 2016-03-01 10 42 30 2014-09-01 85 85 30 2015-05-01 34 119 30 2015-08-01 29 148 30 2016-0

#### <span style="font-size: 12pt; font-family: Arial, sans-serif; color: rgb(102, 102, 102); background-color: transparent; font-weight: 400; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">Chapter 04 - Subqueries.sql</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">---------------------------------------------------------------------</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">-- Substitution Error in a Subquery Column Name</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">---------------------------------------------------------------------</span>

<span style="font-size: 12pt; font-family: Arial, sans-serif; color: rgb(102, 102, 102); background-color: transparent; font-weight: 400; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">The following SQL code is displaying a table "MyShippers" being created with columns names shipper_id, companyname, and phone. This code is specifically trying to portray a common mistake in SQL of misrefrencing a column due to no alias being used. This can be seen in the first query where it's unclear if shipperid should be referenced from Sales.[Order] or Sales.MyShipper. However this is amended in the 2nd and 3rd queries with aliases</span>

In [ ]:
USE Northwinds2022TSQLV7
DROP TABLE IF EXISTS Sales.MyShippers;

CREATE TABLE Sales.MyShippers
(
  shipper_id  INT          NOT NULL,
  companyname NVARCHAR(40) NOT NULL,
  phone       NVARCHAR(24) NOT NULL,
  CONSTRAINT PK_MyShippers PRIMARY KEY(shipper_id)
);

INSERT INTO Sales.MyShippers(shipper_id, companyname, phone)
  VALUES(1, N'Shipper GVSUA', N'(503) 555-0137'),
	      (2, N'Shipper ETYNR', N'(425) 555-0136'),
				(3, N'Shipper ZHISN', N'(415) 555-0138');

-- Shippers who shipped orders to customer 43

-- Bug
SELECT shipper_id, companyname
FROM Sales.MyShippers
WHERE shipper_id IN
 (SELECT shipperid
   FROM Sales.[Order]
   WHERE CustomerId = 43);


-- The safe way using aliases, bug identified
SELECT shipper_id, companyname
FROM Sales.MyShippers
WHERE shipper_id IN
  (SELECT O.ShipperId
   FROM Sales.[Order] AS O
   WHERE O.CustomerId = 43);
GO

-- Bug corrected
SELECT shipper_id, companyname
FROM Sales.MyShippers
WHERE shipper_id IN
  (SELECT O.ShipperId
   FROM Sales.[Order] AS O
   WHERE O.CustomerId = 43);

-- Cleanup
DROP TABLE IF EXISTS Sales.MyShippers;

#### <span style="font-size: 12pt; font-family: Arial, sans-serif; color: rgb(102, 102, 102); background-color: transparent; font-weight: 400; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">Chapter 05 - Table Expressions - Exercises.sql</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">-- 1</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">-- The following query attempts to filter orders placed on the last day of the year.</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 128, 0); background-color: transparent; font-weight: 700; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">USE</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">TSQLV4;</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 128, 0); background-color: transparent; font-weight: 700; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">GO</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 128, 0); background-color: transparent; font-weight: 700; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">SELECT</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">orderid,</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">orderdate,</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">custid,</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">empid,</span>

  <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 255); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">DATEFROMPARTS</span><span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">(</span><span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 255); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">YEAR</span><span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">(orderdate),</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(102, 102, 102); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">12</span><span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">,</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(102, 102, 102); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">31</span><span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">)</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 128, 0); background-color: transparent; font-weight: 700; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">AS</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">endofyear</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 128, 0); background-color: transparent; font-weight: 700; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">FROM</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">Sales.Orders</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 128, 0); background-color: transparent; font-weight: 700; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">WHERE</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">orderdate</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(102, 102, 102); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">&lt;&gt;</span> <span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">endofyear;</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); background-color: transparent; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;"><br></span>

<span style="font-size:11pt;font-family:Arial,sans-serif;color:#408080;background-color:transparent;font-weight:400;font-style:italic;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">-- When you try to run this query you get the following error.</span>

<span style="font-size:11pt;font-family:Arial,sans-serif;color:#408080;background-color:transparent;font-weight:400;font-style:italic;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">/*</span>

<span style="font-size:11pt;font-family:Arial,sans-serif;color:#408080;background-color:transparent;font-weight:400;font-style:italic;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Msg 207, Level 16, State 1, Line 233</span>

<span style="font-size:11pt;font-family:Arial,sans-serif;color:#408080;background-color:transparent;font-weight:400;font-style:italic;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Invalid column name 'endofyear'.</span>

<span style="font-size:11pt;font-family:Arial,sans-serif;color:#408080;background-color:transparent;font-weight:400;font-style:italic;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">*/</span>

<span style="font-size:11pt;font-family:Arial,sans-serif;color:#408080;background-color:transparent;font-weight:400;font-style:italic;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">-- Explain what the problem is and suggest a valid solution.</span>

This problem is occuring because you can't reference endofyear in the WHERE clause as it gets evaluated before the SELECT clause which is where the alias is defined.

To solve this you can repeat the code in the SELECT clause defining endofyear in the WHERE clause:

In [4]:
USE Northwinds2022TSQLV7;
GO

SELECT orderid, orderdate, CustomerId, EmployeeId,
  DATEFROMPARTS(YEAR(orderdate), 12, 31) AS endofyear
FROM Sales.[Order]
WHERE orderdate <> DATEFROMPARTS(YEAR(orderdate), 12, 31);

Commands completed successfully.

(827 rows affected)

Total execution time: 00:00:00.011

orderid orderdate CustomerId EmployeeId endofyear 10248 2014-07-04 85 5 2014-12-31 10249 2014-07-05 79 6 2014-12-31 10250 2014-07-08 34 4 2014-12-31 10251 2014-07-08 84 3 2014-12-31 10252 2014-07-09 76 4 2014-12-31 10253 2014-07-10 34 3 2014-12-31 10254 2014-07-11 14 5 2014-12-31 10255 2014-07-12 68 9 2014-12-31 10256 2014-07-15 88 3 2014-12-31 10257 2014-07-16 35 4 2014-12-31 10258 2014-07-17 20 1 2014-12-31 10259 2014-07-18 13 4 2014-12-31 10260 2014-07-19 56 4 2014-12-31 10261 2014-07-19 61 4 2014-12-31 10262 2014-07-22 65 8 2014-12-31 10263 2014-07-23 20 9 2014-12-31 10264 2014-07-24 24 6 2014-12-31 10265 2014-07-25 7 2 2014-12-31 10266 2014-07-26 87 3 2014-12-31 10267 2014-07-29 25 4 2014-12-31 10268 2014-07-30 33 8 2014-12-31 10269 2014-07-31 89 5 2014-12-31 10270 2014-08-01 87 1 2014-12-31 10271 2014-08-01 75 6 2014-12-31 10272 2014-08-02 65 6 2014-12-31 10273 2014-08-05 63 3 2014-12-31 10274 2014-08-06 85 6 2014-12-31 10275 2014-08-07 49 1 2014-12-31 10276 2014-08-08 80 8 2014-12-31 10277 2014-08-09 52 2 2014-12-31 10278 2014-08-12 5 8 2014-12-31 10279 2014-08-13 44 8 2014-12-31 10280 2014-08-14 5 2 2014-12-31 10281 2014-08-14 69 4 2014-12-31 10282 2014-08-15 69 4 2014-12-31 10283 2014-08-16 46 3 2014-12-31 10284 2014-08-19 44 4 2014-12-31 10285 2014-08-20 63 1 2014-12-31 10286 2014-08-21 63 8 2014-12-31 10287 2014-08-22 67 8 2014-12-31 10288 2014-08-23 66 4 2014-12-31 10289 2014-08-26 11 7 2014-12-31 10290 2014-08-27 15 8 2014-12-31 10291 2014-08-27 61 6 2014-12-31 10292 2014-08-28 81 1 2014-12-31 10293 2014-08-29 80 1 2014-12-31 10294 2014-08-30 65 4 2014-12-31 10295 2014-09-02 85 2 2014-12-31 10296 2014-09-03 46 6 2014-12-31 10297 2014-09-04 7 5 2014-12-31 10298 2014-09-05 37 6 2014-12-31 10299 2014-09-06 67 4 2014-12-31 10300 2014-09-09 49 2 2014-12-31 10301 2014-09-09 86 8 2014-12-31 10302 2014-09-10 76 4 2014-12-31 10303 2014-09-11 30 7 2014-12-31 10304 2014-09-12 80 1 2014-12-31 10305 2014-09-13 55 8 2014-12-31 10306 2014-09-16 69 1 2014-12-31 10307 2014-09-17 48 2 2014-12-31 10308 2014-09-18 2 7 2014-12-31 10309 2014-09-19 37 3 2014-12-31 10310 2014-09-20 77 8 2014-12-31 10311 2014-09-20 18 1 2014-12-31 10312 2014-09-23 86 2 2014-12-31 10313 2014-09-24 63 2 2014-12-31 10314 2014-09-25 65 1 2014-12-31 10315 2014-09-26 38 4 2014-12-31 10316 2014-09-27 65 1 2014-12-31 10317 2014-09-30 48 6 2014-12-31 10318 2014-10-01 38 8 2014-12-31 10319 2014-10-02 80 7 2014-12-31 10320 2014-10-03 87 5 2014-12-31 10321 2014-10-03 38 3 2014-12-31 10322 2014-10-04 58 7 2014-12-31 10323 2014-10-07 39 4 2014-12-31 10324 2014-10-08 71 9 2014-12-31 10325 2014-10-09 39 1 2014-12-31 10326 2014-10-10 8 4 2014-12-31 10327 2014-10-11 24 2 2014-12-31 10328 2014-10-14 28 4 2014-12-31 10329 2014-10-15 75 4 2014-12-31 10330 2014-10-16 46 3 2014-12-31 10331 2014-10-16 9 9 2014-12-31 10332 2014-10-17 51 3 2014-12-31 10333 2014-10-18 87 5 2014-12-31 10334 2014-10-21 84 8 2014-12-31 10335 2014-10-22 37 7 2014-12-31 10336 2014-10-23 60 7 2014-12-31 10337 2014-10-24 25 4 2014-12-31 10338 2014-10-25 55 4 2014-12-31 10339 2014-10-28 51 2 2014-12-31 10340 2014-10-29 9 1 2014-12-31 10341 2014-10-29 73 7 2014-12-31 10342 2014-10-30 25 4 2014-12-31 10343 2014-10-31 44 4 2014-12-31 10344 2014-11-01 89 4 2014-12-31 10345 2014-11-04 63 2 2014-12-31 10346 2014-11-05 65 3 2014-12-31 10347 2014-11-06 21 4 2014-12-31 10348 2014-11-07 86 4 2014-12-31 10349 2014-11-08 75 7 2014-12-31 10350 2014-11-11 41 6 2014-12-31 10351 2014-11-11 20 1 2014-12-31 10352 2014-11-12 28 3 2014-12-31 10353 2014-11-13 59 7 2014-12-31 10354 2014-11-14 58 8 2014-12-31 10355 2014-11-15 4 6 2014-12-31 10356 2014-11-18 86 6 2014-12-31 10357 2014-11-19 46 1 2014-12-31 10358 2014-11-20 41 5 2014-12-31 10359 2014-11-21 72 5 2014-12-31 10360 2014-11-22 7 4 2014-12-31 10361 2014-11-22 63 1 2014-12-31 10362 2014-11-25 9 3 2014-12-31 10363 2014-11-26 17 4 2014-12-31 10364 2014-11-26 19 1 2014-12-31 10365 2014-11-27 3 3 2014-12-31 10366 2014-11-28 29 8 2014-12-31 10367 2014-11-28 83 7 2014-12-31 10

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">-- 6-2</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">-- Using the CROSS APPLY operator</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">-- and the function you created in exercise 6-1,</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">-- return, for each supplier, the two most expensive products</span>

Below is 6-1 where its an inline function that accepts a supplier id and a requested number of products and returns n products with their highest unit prices

In [7]:
USE Northwinds2022TSQLV7
IF OBJECT_ID('Production.TopProducts') IS NOT NULL
   DROP FUNCTION Production.TopProducts;
GO 
CREATE FUNCTION Production.TopProducts
   (@supid AS INT, @n AS INT)
      RETURNS TABLE 
AS 
RETURN 
SELECT TOP (@n) productid, productname, unitprice
    FROM Production.Product   
    WHERE supplierid = @supid   
    ORDER BY unitprice DESC; 
GO

Commands completed successfully.

Commands completed successfully.

Total execution time: 00:00:00.009

Here the query fetches information about suppliers and, for each supplier, retrieves their top 2 products based on unit price. 

Passing a 2 into the function returns the **top 2** highest products and the Cross Apply essntially join s each supplier with their top products and this can be seen in the resulting table below

In [10]:
USE Northwinds2022TSQLV7
IF OBJECT_ID('Production.TopProducts') IS NOT NULL
   DROP FUNCTION Production.TopProducts;
GO 
CREATE FUNCTION Production.TopProducts
   (@supid AS INT, @n AS INT)
      RETURNS TABLE 
AS 
RETURN 
SELECT TOP (@n) productid, productname, unitprice
    FROM Production.Product   
    WHERE supplierid = @supid   
    ORDER BY unitprice DESC; 
GO

SELECT S.supplierid,
       S.SupplierCompanyName as companyname,
       P.productid,
       P.productname,
       P.unitprice
FROM Production.Supplier AS S
    CROSS APPLY Production.TopProducts(S.supplierid, 2) AS P;

Commands completed successfully.

Commands completed successfully.

(55 rows affected)

Total execution time: 00:00:00.012

supplierid companyname productid productname unitprice 1 Supplier SWRXU 2 Product RECZE 19.00 1 Supplier SWRXU 1 Product HHYDP 18.00 2 Supplier VHQZD 4 Product KSBRM 22.00 2 Supplier VHQZD 5 Product EPEIM 21.35 3 Supplier STUAZ 8 Product WVJFP 40.00 3 Supplier STUAZ 7 Product HMLNI 30.00 4 Supplier QOVFD 9 Product AOZBW 97.00 4 Supplier QOVFD 10 Product YHXGE 31.00 5 Supplier EQPNC 12 Product OSFNS 38.00 5 Supplier EQPNC 11 Product QMVUN 21.00 6 Supplier QWUSF 14 Product PWCJB 23.25 6 Supplier QWUSF 15 Product KSZOI 15.50 7 Supplier GQRCV 18 Product CKEDC 62.50 7 Supplier GQRCV 63 Product ICKNK 43.90 8 Supplier BWGYE 20 Product QHFFP 81.00 8 Supplier BWGYE 68 Product TBTBL 12.50 9 Supplier QQYEU 22 Product CPHFY 21.00 9 Supplier QQYEU 23 Product JLUDZ 9.00 10 Supplier UNAHG 24 Product QOGNU 4.50 11 Supplier ZPYVS 27 Product SMIOH 43.90 11 Supplier ZPYVS 26 Product HLGZA 31.23 12 Supplier SVIYA 29 Product VJXYN 123.79 12 Supplier SVIYA 28 Product OFBNT 45.60 13 Supplier TEGSC 30 Product LYERX 25.89 14 Supplier KEREV 72 Product GEEOO 34.80 14 Supplier KEREV 32 Product NUNAW 32.00 15 Supplier NZLIF 69 Product COAXA 36.00 15 Supplier NZLIF 71 Product MYMOI 21.50 16 Supplier UHZRG 35 Product NEVTJ 18.00 16 Supplier UHZRG 67 Product XLXQF 14.00 17 Supplier QZGUF 37 Product EVFFA 26.00 17 Supplier QZGUF 36 Product GMKIJ 19.00 18 Supplier LVJUA 38 Product QDOMO 263.50 18 Supplier LVJUA 39 Product LSOFL 18.00 19 Supplier JDNUG 40 Product YZIXQ 18.40 19 Supplier JDNUG 41 Product TTEEX 9.65 20 Supplier CIYNM 43 Product ZZZHumanResources 46.00 20 Supplier CIYNM 44 Product VJIEO 19.45 21 Supplier XOXZA 46 Product CBRRL 12.00 21 Supplier XOXZA 45 Product AQOKR 9.50 22 Supplier FNUXM 48 Product MYNXN 12.75 22 Supplier FNUXM 47 Product EZZPR 9.50 23 Supplier ELCRN 49 Product FPYPN 20.00 23 Supplier ELCRN 76 Product JYGFE 18.00 24 Supplier JNNES 51 Product APITJ 53.00 24 Supplier JNNES 53 Product BKGEA 32.80 25 Supplier ERVYZ 55 Product YYWRT 24.00 25 Supplier ERVYZ 54 Product QAQRL 7.45 26 Supplier ZWZDM 56 Product VKCMF 38.00 26 Supplier ZWZDM 57 Product OVLQI 19.50 27 Supplier ZRYDZ 58 Product ACRVI 13.25 28 Supplier OAVQT 59 Product UKXRI 55.00 28 Supplier OAVQT 60 Product WHBYK 34.00 29 Supplier OGLRK 62 Product WUXYK 49.30 29 Supplier OGLRK 61 Product XYZPE 28.50

#### <span style="font-size: 12pt; font-family: Arial, sans-serif; color: rgb(102, 102, 102); background-color: transparent; font-weight: 400; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">Chapter 05 - Table Expressions.sql</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">---------------------------------------------------------------------</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">-- Using Arguments</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">---------------------------------------------------------------------</span>

<span style="font-size: 14.6667px; white-space: pre-wrap;">The query counts the distinct customers served by Employee 3 each year based on order dates in the "Sales.Order" table.</span>

In [11]:
USE Northwinds2022TSQLV7
-- Yearly Count of Customers handled by Employee 3
DECLARE @empid AS INT = 3;

SELECT orderyear, COUNT(DISTINCT D.CustomerId) AS numcusts
FROM (SELECT YEAR(orderdate) AS orderyear, O.CustomerId
      FROM Sales.[Order] as O
      WHERE EmployeeId = @empid) AS D
GROUP BY orderyear;
GO

(3 rows affected)

Total execution time: 00:00:00.012

orderyear numcusts 2014 16 2015 46 2016 30

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">---------------------------------------------------------------------</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">-- Nesting</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">---------------------------------------------------------------------</span>

<span style="font-size: 14.6667px; white-space: pre-wrap;">The query calculates the count of distinct customers per year from the "Sales.Order" table, then filters the results to display only the years with more than 70 distinct customers, demonstrating a nested derived table approach and an alternative approach using HAVING clause.</span>

In [12]:
USE Northwinds2022TSQLV7
-- Listing 5-2 Query with Nested Derived Tables
SELECT orderyear, numcusts
FROM (SELECT orderyear, COUNT(DISTINCT D1.CustomerID) AS numcusts
      FROM (SELECT YEAR(orderdate) AS orderyear, O.CustomerId
            FROM Sales.[Order] as O) AS D1
      GROUP BY orderyear) AS D2
WHERE numcusts > 70;

SELECT YEAR(orderdate) AS orderyear, COUNT(DISTINCT O.CustomerId) AS numcusts
FROM Sales.[Order] as O
GROUP BY YEAR(orderdate)
HAVING COUNT(DISTINCT O.CustomerId) > 70;


(2 rows affected)

(2 rows affected)

Total execution time: 00:00:00.014

orderyear numcusts 2015 86 2016 81

orderyear numcusts 2015 86 2016 81

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">---------------------------------------------------------------------</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">-- Views Described</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">---------------------------------------------------------------------</span>

<span style="font-size: 14.6667px; white-space: pre-wrap;">The query creates a view named "USACusts" to select specific customer details for customers from the USA in the "Sales.Customer" table and then retrieves the customer ID and company name from this view.</span>

In [13]:
Use Northwinds2022TSQLV7
---------------------------------------------------------------------
-- Views Described
---------------------------------------------------------------------

-- Creating USACusts View
DROP VIEW IF EXISTS Sales.USACusts;
GO
CREATE VIEW Sales.USACusts
AS

SELECT
  CustomerId, CustomerCompanyName, CustomerContactName, CustomerContactTitle, CustomerAddress,
  CustomerCity, CustomerRegion, CustomerPostalCode, CustomerCountry,CustomerPhoneNumber, CustomerFaxNumber
FROM Sales.Customer
WHERE CustomerCountry = N'USA';
GO

SELECT CustomerId, CustomerCompanyName
FROM Sales.USACusts;
GO


Commands completed successfully.

Commands completed successfully.

(13 rows affected)

Total execution time: 00:00:00.016

CustomerId CustomerCompanyName 32 Customer YSIQX 36 Customer LVJSO 43 Customer UISOJ 45 Customer QXPPT 48 Customer DVFMB 55 Customer KZQZT 65 Customer NYUHS 71 Customer LCOUJ 75 Customer XOJYP 77 Customer LCYBZ 78 Customer NLTYP 82 Customer EYHKM 89 Customer YBQTI

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">---------------------------------------------------------------------</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">-- Inline User Defined Functions</span>

<span style="font-size: 11pt; font-family: Arial, sans-serif; color: rgb(64, 128, 128); background-color: transparent; font-style: italic; font-variant-numeric: normal; font-variant-east-asian: normal; vertical-align: baseline; white-space: pre-wrap;">---------------------------------------------------------------------</span>

<span style="font-size: 14.6667px; white-space: pre-wrap;">The query creates a table-valued function named "GetCustOrders" to retrieve order details for a specified customer ID from the "Sales.Order" table. It then tests the function by retrieving order IDs and customer IDs for customer ID 1, followed by retrieving specific order and product details for customer ID 1. Finally, it cleans up by dropping the function.</span>

In [16]:
-- Creating GetCustOrders function
USE Northwinds2022TSQLV7;
DROP FUNCTION IF EXISTS dbo.GetCustOrders;
GO
CREATE FUNCTION dbo.GetCustOrders
  (@cid AS INT) RETURNS TABLE
AS
RETURN
  SELECT orderid, CustomerId, EmployeeId, orderdate, requireddate,
    ShipToDate, shipperid, freight, ShipToName, ShipToAddress, ShipToCity,
    ShipToRegion, ShipToPostalCode, ShipToCountry
  FROM Sales.[Order]
  WHERE CustomerId = @cid;
GO

-- Test Function
SELECT orderid, CustomerId
FROM dbo.GetCustOrders(1) AS O;

SELECT O.orderid, O.CustomerId, OD.productid, OD.Quantity
FROM dbo.GetCustOrders(1) AS O
  INNER JOIN Sales.OrderDetail AS OD
    ON O.orderid = OD.orderid;
GO

-- Cleanup
DROP FUNCTION IF EXISTS dbo.GetCustOrders;
GO


Commands completed successfully.

Commands completed successfully.

(6 rows affected)

(12 rows affected)

Commands completed successfully.

Total execution time: 00:00:00.018

orderid CustomerId 10643 1 10692 1 10702 1 10835 1 10952 1 11011 1

orderid CustomerId productid Quantity 10643 1 28 15 10643 1 39 21 10643 1 46 2 10692 1 63 20 10702 1 3 6 10702 1 76 15 10835 1 59 15 10835 1 77 2 10952 1 6 16 10952 1 28 2 11011 1 58 40 11011 1 71 20